src/actions/app.cr

```diff
@@ -49,86 +50,174 @@ module DppmRestApi::Actions::App
   # current log data.
   relative_get "/:app_name/logs" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
-    # TODO: upgrade to websocket or output logs to date
+    app_name = context.params.url["app_name"]
+    app = Actions.prefix.new_app app_name
+    stream_type = context.params.query["stream_type"]? || [] of String
+    stream_type = if stream_type.empty?
+                    ["errors", "output"]
+                  else
+                    ["errors", "output"] & stream_type
+                  end
+    if context.params.query["stream"]?
+      raise BadRequest.new "You must use separate requests to stream both STDOUT and STDERR." if stream_type.size != 1
+      context.response.upgrade do |socket|
+        app.get_logs(
+          stream_type.first + ".log",
+          follow: true,
+          lines: context.params.query["lines"]?.try &.to_i
+        ) { |line| socket.puts line }
+      end
+    else
+      JSON.build context.response do |builder|
+        builder.object do
+          builder.field "logs" do
+            builder.object do
+              stream_type.each do |stream|
+                builder.field stream do
+                  builder.array do
+                    app.get_logs stream + ".log", false, context.params.query["lines"]?.try(&.to_i) do |line|
+                      builder.string line
+                    end
+                  end
+                end
+              end
+            end
+          end
+        end
+      end
+    end
+  end
+
+  {% for type in ["error", "output"] %}
+  # Stream the logs for the given application over the websocket connection.
+  relative_ws "/:app_name/logs/{{type.id}}" do |sock, context|
+    app_name = context.params.url["app_name"]
+    raise Unauthorized.new context unless Actions.has_access? context, Access::Read
+    app.get_logs(
+      "{{type.id}}.log",
+      follow: true,
+      lines: context.params.query["lines"]?.try(&.to_i)
+    ) do |line|
+      sock.puts line
+    end
   end
+  {% end %}
+
   # Install the given package
   relative_put "/:app_name" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Create
     # TODO: install the package and return its name
   end
+
+  {% for param, comment in {preserve_database: "database", keep_user_group: "user group"} %}
+  # For when deleting an application, a check of the query parameter on whether
+  # to keep {{comment.id}} for the app.
+  private macro {{param.id}}?
+    if %values = context.params.query["{{param.id}}"]?
+      if %values.empty?
+        # The parameter was specified with no text.
+        true
+      elsif !(["false", "no"] & %values.map(&.downcase)).empty?
+        # {{param.id}}=no is ok too
+        false
+      else
+        # Anything but false and no get assumed to be a truthy value.
+        true
+      end
+    else
+      false
+    end
+  end
+  {% end %}
```

Comment:
Please no macro,  use a method instead like `cast_bool_param(name, context)`, something like this.

Reply:
This already exists -- the code was simply outdated. Thank you for pointing it out.